google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ServerToIDEFileResolver.java

```diff
@@ -0,0 +1,124 @@
+import com.intellij.openapi.vfs.VirtualFileSystem;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiJavaFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.search.FileTypeIndex;
+import com.intellij.psi.search.GlobalSearchScope;
+import com.intellij.util.indexing.FileBasedIndex;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Provides a translation between file names sent from the CDB API and IntelliJ project files in
+ * the local file system.
+ *
+ * <p>This is necessary mostly because StackFrames locations in CDB only contain the package and
+ * file name. The Cloud Debugger API returns StackFrame.SourceLocation.path in the form of
+ *
+ * com/my/package/Class.java
+ *
+ * while IntelliJ offers VirtualFiles whose path is the absolute path from root. e.g.,
+ *
+ * /home/user/workspace/repo/path/com/my/package/Class.java
+ *
+ * <p>These are methods which are Java specific for the cloud debugger. When we add other languages,
+ * some of this may need to be extracted to an extensionpoint.
+ */
+public class ServerToIDEFileResolver {
+  private final Map<Project, Map<String, VirtualFile>> ourLocationMaps =
+      new ConcurrentHashMap<Project, Map<String, VirtualFile>>();
+  private VirtualFileSystem fileSystem =
+      VirtualFileManager.getInstance().getFileSystem(StandardFileSystems.FILE_PROTOCOL);
+
+  public ServerToIDEFileResolver(@NotNull final Project project, boolean forceRefresh) {
+    initializeLocationsForProject(project, forceRefresh);
+  }
+
+  /**
+   * Utility method that returns the full class name for a file.
+   */
+  public static String getCloudPathFromJavaFile(PsiJavaFile javaFile) {
+    return javaFile.getPackageName().replace('.', '/') + "/" + javaFile.getName();
+  }
+
+  /**
+   * Given a possible file path, returns a VirtualFile instance that the IDE can work with.
+   *
+   * <p>This method tries to fetch a file in three ways. First, it uses its full path in the local
+   * file system. Then, it tries the full class name form (com/google/gct/idea/debugger/
+   * CloudDebugProcess.java). Finally, it searches for possible file matches within the project.
+   */
+  public VirtualFile getFileFromPath(@NotNull Project project, @NotNull String path) {
+    // Try the relative full project path.
+    VirtualFile file = fileSystem.findFileByPath(project.getBasePath() + "/" + path);
+    // Try class name with package and class file name.
+    if (file == null) {
+      file = ourLocationMaps.get(project).get(path);
+    }
+    // If we still couldn't find the file, search for possible file name matches and return the
+    // first.
+    // We might want to improve string matching and return more than one possible match.
+    if (file == null) {
+      Collection<VirtualFile> projectJavaFiles = FileBasedIndex.getInstance().getContainingFiles(
```

Comment:
I did that at first, but then understood that I'd have to inject a PsiManager, etc.

It turns out that the recommended way of testing a project with Psi files is by using the mocks provided by IDEA, that I'm using in my tests.